libyul/optimiser/MemoryEscalator.cpp

```diff
@@ -0,0 +1,137 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <libyul/optimiser/MemoryEscalator.h>
+#include <libyul/optimiser/CallGraphGenerator.h>
+#include <libyul/optimiser/FunctionCallFinder.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/optimiser/StackToMemoryMover.h>
+#include <libyul/backends/evm/EVMDialect.h>
+#include <libyul/AsmData.h>
+#include <libyul/CompilabilityChecker.h>
+#include <libyul/Dialect.h>
+#include <libyul/Exceptions.h>
+#include <libyul/Object.h>
+#include <libyul/Utilities.h>
+#include <libsolutil/Algorithms.h>
+#include <libsolutil/CommonData.h>
+#include <list>
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+
+void MemoryEscalator::run(OptimiserStepContext& _context, Object& _object, bool _optimizeStackAllocation)
+{
+	if (!_object.code)
+		return;
+
+	// Determine which variables need to be moved.
+	auto functionStackErrorInfo = CompilabilityChecker::run(_context.dialect, _object, _optimizeStackAllocation);
+	if (functionStackErrorInfo.empty())
+		return;
+
+	// Find the literal argument of the ``memoryinit`` call, if there is a unique such call, otherwise abort.
+	yulAssert(dynamic_cast<EVMDialect const*>(&_context.dialect), "MemoryEscalator can only be run on EVMDialect objects.");
+	Literal* memoryInitLiteral = nullptr;
+	if (
+		auto memoryInits = FunctionCallFinder::run(*_object.code, "memoryinit"_yulstring);
```

Comment:
I'm wondering - wouldn't it be better to search `memoryinit` among the first statements? Then we can be sure it is not e.g. hidden inside a branch

Reply:
Hm, maybe... I did that before when it still just looked for an ``mstore(0x40, ...)``... I'm not particularly happy with searching among the first statements... but maybe it's the best we can do, I'll think about whether I can come up with something better - might make the ``FunctionCallFinder`` obsolete again, though, I only use it for this here :-).